Issue-52: Implement Friendship

chrispinkney opened the pull request:
## Issue
Closes #52

## Type of Change
<!-- bug fix, feature, documentation, UI, etc. -->

- [ ] **Bugfix**: Change which fixes an issue
- [x] **New Feature**: Change which adds functionality
- [ ] **Documentation Update**: Change which improves documentation
- [ ] **UI**: Change which improves UI

## Description
This PR adds the ability to implement friendships between users. Friendships can be one of three status': ` 'ACCEPT' | 'REJECT' | 'PENDING' | 'SENT'`.

- Users can get their friends/friend requests via: `GET {{BASE_URL}}/auth/friends`
- Users can send a friend request via: `POST {{BASE_URL}}/auth/friends/send`:
```json
{
    "username": "chris"
}
```
- Users can update a friend request via: `POST {{BASE_URL}}/auth/friends/update`:
```json
{
    "username": "chris",
    "status": "ACCEPT"
}
```
- Users can reject a friend request (and delete a friendship) via: `POST {{BASE_URL}}/auth/friends/update`:
```json
{
    "username": "chris",
    "status": "REJECT"
}
```


## Testing the PR
Create several friends and try to create friend requests between them. Run the GET request during each step to see the potential status' and reactions from the backend.

## Checklist
<!-- Before submitting a PR, address each item -->

- [ ] **Quality**: This PR builds and passes the respective npm run test and works locally
- [ ] **Tests**: This PR includes thorough tests
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [ ] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

Thinbishop commented on api/src/users/users.controller.ts:
```diff
@@ -126,6 +148,34 @@ export class UsersController {
+	@UseInterceptors(RemoveFieldsInterceptor)
+	@UseGuards(JwtAuthGuard)
+	@Post('/friends/send')
+	@HttpCode(HttpStatus.OK)
```
Should this be `201: Created`? I.e. you could just remove the line and `POST` should default to 201.

chrispinkney replied:
Does it default to 201? I thought it was 200, either way sure!

Thinbishop reviewed (commented):
Would it be useful to also have a deletion for friendships? This could be done both when there is a friend request or they are already friends. From what I can see, it should work the same way.

Other than that, it looks great! Excited to have this implemented!

Thinbishop reviewed (commented):
A couple things I noticed when testing this out:
- When a friend request is pending, it shows up as pending for both users, so the frontend will not be able to distinguish which user can accept/reject the friend request.
- When sending a request to someone you've already sent a request to, the server will respond with `{statusCode: 409, message: "Email already in use"}` - may need to change the message here if possible.
- If a friendship has been rejected, it will still be contained within the friends list (using GET `/auth/friends/send`), with a status of "REJECT". At that point, it's impossible to reinitiate that friendship, as it responds with 409 (see point above)

chrispinkney commented:
> Would it be useful to also have a deletion for friendships? This could be done both when there is a friend request or they are already friends. From what I can see, it should work the same way.
> 
> Other than that, it looks great! Excited to have this implemented!


Done, but instead of making a whole delete route I simply tied it to reject since rejecting a friend request will have the same behaviour. I updated the PR description to reflect this change.


> A couple things I noticed when testing this out:
> 
>     * When a friend request is pending, it shows up as pending for both users, so the frontend will not be able to distinguish which user can accept/reject the friend request.
> 
>     * When sending a request to someone you've already sent a request to, the server will respond with `{statusCode: 409, message: "Email already in use"}` - may need to change the message here if possible.
> 
>     * If a friendship has been rejected, it will still be contained within the friends list (using GET `/auth/friends/send`), with a status of `"REJECT"`. At that point, it's impossible to reinitiate that friendship, as it responds with 409 (see point above)
> 
> 
> Edit: I can reinitiate the friendship from a `"REJECT"` state if I POST to `/update` with an `"ACCEPT"` status from either user.


1. I added a new state suchthat user1's state will be 'SENT' and user2's state will be 'PENDING'. If a state if pending, show the accept/reject buttons on the UI. If the state if 'SENT' show only a cancel button to cancel the friend request (user1 cannot accept a friend request from user2, only delete the request.)
2. I added a new handle for that: `throw new BadRequestException('Friendship already exists');`
3. See above comment about this no longer being an issue. Originally I wanted to persist everything and simply toggle it back and forth but I think this is a better solution. 4 should not be an issue with this change either.

pyvelkov reviewed (approved):
Tested and working. I agree with @chrispinkney about the "rejected" status being the way to delete users. It is going to be the same anyways.
Good job Chris!
